GH-38701: [C++][FS][Azure] Implement `DeleteDirContents()`

kou opened the pull request:
### Rationale for this change

`DeleteDirContents()` deletes the given directory contents recursively like other filesystem implementations.

Azure file system treats the following cases as root directory:
* Empty container
* Empty path

### What changes are included in this PR?

List and delete approach is used with/without hierarchical namespace support. Because Azure doesn't provide "DeleteDirContents" API with hierarchical namespace support. We may want to use delete the given directory and create an empty approach instead of list and delete approach for performance but it may not be acceptable with some use cases. 

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.
* Closes: #38701

kou commented:
@Tom-Newton @felipecrv You may want to review this.

felipecrv commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -970,6 +970,78 @@ class AzureFileSystem::Impl {
+    try {
+      auto list_response = container_client.ListBlobs(options);
+      if (!missing_dir_ok && list_response.Blobs.empty()) {
+        return Status::IOError("Specified directory doesn't exist: ", location.path, ": ",
+                               container_client.GetUrl());
```
Shouldn't this be a `PathNotFound` error?

kou replied:
Good catch!

felipecrv commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -970,6 +970,78 @@ class AzureFileSystem::Impl {
+        return Status::IOError("Specified directory doesn't exist: ", location.path, ": ",
+                               container_client.GetUrl());
+      }
+      while (list_response.HasPage() && !list_response.Blobs.empty()) {
```
This could be me being overly cautious, but maybe we shouldn't stop when `Blobs.empty()`, but instead run `continue;` and evaluate `MoveToNextPage()` again?

felipecrv replied:
Another thing: instead of `while`, you can use `for (; list_response.HasPage(); list_response.MoveToNextPage()) {` here.

kou replied:
Both of `HasPage()` and `!Blobs.empty()` checks are needed to avoid empty `SubmitBatch()` request.
If there are no blobs under the given path, `ListBlobs()` returns `HasPage() == true` and `Blobs.empty() == true` response.

I'll use `for` as you suggested. I also found an azure-sdk-for-cpp document that uses `for`: https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/storage/MigrationGuide.md#listing-blobs-in-a-container

felipecrv replied:
> ListBlobs() returns HasPage() == true and Blobs.empty() == true response.

Is there a chance of that not being the last page? I see that you now only `continue;` instead of breaking the entire loop which I think is the more robust thing to do. 👍

kou replied:
(Hmm. I think that the case (empty blob but it's not the last response) isn't happen but I'm not sure because Azure Blob Storage is a proprietary product...)

felipecrv commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -970,6 +970,78 @@ class AzureFileSystem::Impl {
+            Azure::Storage::Blobs::Models::DeleteBlobResult>>
+            deferred_responses;
+        for (const auto& blob_item : list_response.Blobs) {
+          deferred_responses.push_back(batch.DeleteBlob(blob_item.Name));
```
Skip when the `blob_item.Name` matches the `options.Prefix` because that could be the empty-dir marker blob (an empty blob ending with a `/`).

kou replied:
I think that we should do it in https://github.com/apache/arrow/issues/38772 .
We have a test for this case. So we can detect the case when we work on https://github.com/apache/arrow/issues/38772 .

felipecrv commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -1017,69 +1089,67 @@ class AzureFileSystem::Impl {
-      // size of the body for a batch request can't exceed 4 MB.
-      const int32_t kNumMaxRequestsInBatch = 256;
-      options.PageSizeHint = kNumMaxRequestsInBatch;
+      return DeleteDirContentsWihtoutHierarchicalNamespace(location, true);
```
`/*missing_dir_ok=*/true`

kou replied:
OK. I'll add the comment.

felipecrv commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -1017,69 +1089,67 @@ class AzureFileSystem::Impl {
+      return internal::InvalidDeleteDirContents(location.all);
+    }
+    if (location.path.empty()) {
+      return internal::InvalidDeleteDirContents(location.all);
```
Shouldn't this be interpreted as deleting the contents of the container?

kou replied:
We have a special API for this case: `DeleteRootDirContents()`

https://github.com/apache/arrow/blob/63353baf1cda1d1fc7bb614ce01558c12990e073/cpp/src/arrow/filesystem/filesystem.h#L233-L249

So I think that we should return an error here.

felipecrv replied:
But what is the root dir? Doesn't that mean *all the containers*?

`s3fs` maps the root-dir concept to *all containers*:
https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/s3fs.cc#L2762

Refusing to delete contents of a container can be very limiting as `CreateDir(container/)` will create a container.

kou replied:
Can we clarify this discussion?

What should be "root dir"?

1. ` `: Empty location (no container and no path)
2. `container` and `container/`: Container only
3. `container/path`: Container and path

My opinion and the current implementation:

1. Root dir
2. Root dir
3. Not root dir

FYI: s3fs:

1. ` `: Empty path (no bucket and no key): Root dir
2. `bucket` and `bucket/`: Bucket only: Not root dir
3. `bucket/key`: Bucket and key: Not root dir


@Tom-Newton What do you think about this as an Azure Blob Storage user? I'm not an Azure Blob Storage user.

Tom-Newton replied:
I would consider the storage account the root dir. So that means only  `1` is the root dir. I believe this is how the fsspec implementation and the original arrow azurefs PR did it. 

I think when comparing azure to s3, the Azure storage account is analogous to the S3 bucket and the Azure container is an extra level that S3 doesn't have. I think `1` doesn't really reference any storage on S3.

kou replied:
OK. I'll do it.

felipecrv replied:
> I think when comparing azure to s3, the Azure storage account is analogous to the S3 bucket and the Azure container is an extra level that S3 doesn't have. I think 1 doesn't really reference any storage on S3.

Azure Containers are S3 Buckets. Azure Blob Names are S3 keys. `s3fs` has an option [1] that forbids/allows creation of new buckets for when the user asks for the creation of a `/newbucket` directory [2].

[1] https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/s3fs.cc#L1820
[2] https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/s3fs.cc#L2672

felipecrv replied:
There is also an `allow_bucket_deletion` [1] in s3fs. An analogous to that, but for Azure (let's say `allow_container_deletion`) would be relevant in this PR.

[1] https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/s3fs.cc#L2718

Tom-Newton replied:
Probably I need to review the PR properly to understand the context. I don't feel strongly.

kou replied:
The `allow_bucket_deletion` option is for `DeleteDir()` not `DeleteDirContents()`.
And this PR is for `DeleteDirContents()`.
So it should be discussed separately.
I've opened a new issue for it: #38999
We should discuss it further on the issue not here.

Tom-Newton replied:
Overall I'm happy with how this is currently implemented. 

For the sake of argument though... I think `allow_container_deletion` is relevant to `DeleteDirContents` in the case of `location.container.empty()`. I would expect the behaviour in this case to be to delete all containers in the storage account. However, I'm happy to stick with just returning an error because it seems quite dangerous.

kou commented:
Updated:
* Use `PathNotFound()`
* Use `for`
* Add an argument name comment
* Skip a test on macOS that has a problem with Azurite

kou commented:
Updated:
* Accept `DeleteDir("container")`

kou commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -970,6 +970,79 @@ class AzureFileSystem::Impl {
+    auto container_client =
+        blob_service_client_->GetBlobContainerClient(location.container);
+    Azure::Storage::Blobs::ListBlobsOptions options;
+    options.Prefix = internal::EnsureTrailingSlash(location.path);
```
`EnsureTrailingSlash()` does nothing for empty string:
https://github.com/apache/arrow/blob/d66780d90c6faf7bf051cbf12b4b68dff098bd54/cpp/src/arrow/filesystem/path_util.cc#L151-L159

Should we add a (needless) special condition such as the following for readability?

```cpp
if (!location.path.empty()) {
  options.Prefix = internal::EnsureTrailingSlash(location.path);
}
```

Tom-Newton replied:
Maybe just leave a comment saying that `EnsureTrailingSlash` doesn't alter empty string.

felipecrv replied:
Wow. From the name, I though it **ensured** a trailing slash on the output. 🥲

I like this because `Prefix` is a nullable, so not doing anything to it makes even more sense than setting it to a `""` value.

```cpp
if (!location.path.empty()) {
  options.Prefix = internal::EnsureTrailingSlash(location.path);
}
```

kou replied:
OK. I'll use the `if (!location.path.empty())`.

Tom-Newton commented on cpp/src/arrow/filesystem/azurefs_test.cc:
```diff
@@ -666,6 +666,119 @@ TEST_F(AzuriteFileSystemTest, DeleteDirUri) {
+  GTEST_SKIP() << "This test fails by an Azurite problem: "
+                  "https://github.com/Azure/Azurite/pull/2302";
+#endif
+  const auto container_path = RandomContainerName();
+  const auto directory_path =
+      internal::ConcatAbstractPath(container_path, RandomDirectoryName());
+  const auto sub_directory_path = internal::ConcatAbstractPath(directory_path, "new-sub");
+  const auto sub_blob_path = internal::ConcatAbstractPath(sub_directory_path, "sub.txt");
+  const auto top_blob_path = internal::ConcatAbstractPath(directory_path, "top.txt");
+  ASSERT_OK(fs_->CreateDir(sub_directory_path, true));
+  ASSERT_OK_AND_ASSIGN(auto output, fs_->OpenOutputStream(sub_blob_path));
+  ASSERT_OK(output->Write(std::string_view("sub")));
+  ASSERT_OK(output->Close());
+  ASSERT_OK_AND_ASSIGN(output, fs_->OpenOutputStream(top_blob_path));
+  ASSERT_OK(output->Write(std::string_view("top")));
+  ASSERT_OK(output->Close());
```
Could all of this be extracted to a member function of `AzureFileSystemTest`? It looks like this has been copy pasted in 3 of the tests.

kou replied:
OK. I'll extract them as a method.

Tom-Newton reviewed (approved):
I'm happy. Thanks

kou commented:
Updated:
* Rebased on main
* Don't set `ListBlobOptions.Prefix` when `location.path` is empty
* Extract common test data creation codes as a method

I'll merge this once CI is green.
